tests/helpers/deviceappproxy/apduabstract.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Dict, Optional, cast, NewType, Tuple, AnyStr
+from dataclasses import dataclass, field
+
+
+@dataclass
+class CApdu:
+    class Type:
+        IN: int = 0
+        OUT: int = 1
+        INOUT: int = 2
+    data: List[bytes]
+    cla: AnyStr = field(default="00")
+    ins: AnyStr = field(default="00")
+    p1: AnyStr = field(default="00")
+    p2: AnyStr = field(default="00")
+    lc: AnyStr = field(default="00")
+    le: AnyStr = field(default="00")
+    typ: Type = field(default=Type.INOUT)
+
+
+ApduDict = NewType("ApduDict", Dict[str, CApdu])
+
+
+class ApduSet:
+    """Collects a set of CApdu objects and provides their raw version, ready
+    to be sent to a device/an app"
+    """
+    _apdus: ApduDict = {}
+
+    def __init__(self, apdus: Optional[ApduDict] = None, max_lc: int = 255) -> None:
+        self.apdus = apdus
+        self.max_lc = max_lc
+
+    @staticmethod
+    def _bytes(data: AnyStr) -> bytes:
+        if isinstance(data, (bytes, bytearray)):
+            return data
+        if isinstance(data, int):
+            return bytes([cast(int, data)])
+        if isinstance(data, str):
+            return bytes.fromhex(data)
+        raise TypeError(f"{data} cannot be converted to bytes")
+
+    @property
+    def apdus(self) -> Optional[ApduDict]:
+        return ApduSet._apdus if len(ApduSet._apdus.keys()) > 0 else None
+
+    @apdus.setter
+    def apdus(self, new_apdus: ApduDict, overwrite: bool = False) -> None:
+        """Sets a new CApsu internal dictionary if it wasn't set at instanciation time,
+        unless overwrite is True."""
+        if not self.apdus or overwrite is True:
+            if not isinstance(new_apdus, dict):
+                raise ValueError("Attribute newApdus must be a dictionary containing CApdu "
+                                 "instances as values")
+            ApduSet._apdus = new_apdus
+
+    def apdu(self, name: str,
+             p1: Optional[AnyStr] = None,
+             p2: Optional[AnyStr] = None,
+             data: Optional[List[AnyStr]] = None,
+             le: Optional[AnyStr] = None) -> Tuple[bytes, List[Optional[bytes]]]:
+        """Returns the raw bytes for the C-APDU header requested by name.
+        """
+
+        def _bytesbuf(apdu: CApdu, apdu_keys: List[str]) -> bytes:
+            """Concatenates all @apdu attributes whose names are provided in @apdu_keys,
+            into a single byte buffer."""
+            return b''.join(self._bytes(getattr(apdu, k)) if k in apdu.__dict__ else self._bytes(k) for k in apdu_keys)
+
+        if not self.apdus:
+            raise ValueError("ApduSet object is empty! Provide an ApduDict either at instanciation"
+                             " or with the 'apdus' attribute.")
+        if name not in self.apdus:
+            raise KeyError(f"{name} APDU is not supported by this instance")
```

Comment:
I don't see any reason to burden the code with checks on self.apdus.

This is a test framework, so this seems a bit too much. And anyway your code below will raise anyway a ValueError if self.apdus is not set, and a KeyError if name is not a key.

Reply:
This goes for all checks you added in this file and other test classes.

Reply:
This is my embedded C developer skin that shows through :-)

These checks were thought of as explicit hints to ppl that will reuse this class to add support for other apps than BTC. Kind of like a usage manual. I won't remove them now (there are many as you saw) but I'll leave it to the next maintainers (@grydz, @TamtamHero) of this framework to decide of their fate.

Reply:
> This is my "validate inputs before access" embedded C developer skin showing through 😄

Yeah I can tell haha
Relax, it's python, you can give up on your dreams to have proper types :D